[mlir] Attempt to resolve edge cases in PassPipeline textual format

christopherbate opened the pull request:
This commit makes the following changes:

1. Previously certain pipeline options could cause the options parser to
   get stuck in an an infinite loop. An example is:

   ```
   mlir-opt %s -verify-each=false -pass-pipeline='builtin.module(func.func(test-options-super-pass{list={list=1,2},{list=3,4}}))''
   ```

   In this example, the 'list' option of the `test-options-super-pass`
   is itself a pass options specification (this capability was added in
   https://github.com/llvm/llvm-project/issues/101118).

   However, while the textual format allows `ListOption<int>` to be given
   as `list=1,2,3`, it did not allow the same format for
   `ListOption<T>` when T is a subclass of `PassOptions` without extra
   enclosing `{....}`. Lack of enclosing `{...}` would cause the infinite
   looping in the parser.

   This change resolves the parser bug and also allows omitting the
   outer `{...}` for `ListOption`-of-options.

2. Previously, if you specified a default list value for your
   `ListOption`, e.g. `ListOption<int> opt{*this, "list", llvm::cl::list_init({1,2,3})}`,
   it would be impossible to override that default value of `{1,2,3}` with
   an *empty* list on the command line, since `my-pass{list=}` was not allowed.

   This was not allowed because of ambiguous handling of lists-of-strings
   (no literal marker is currently required).

   This change makes it explicit in the ListOption construction that we
   would like to treat all ListOption as having a default value of "empty"
   unless otherwise specified (e.g. using `llvm::list_init`).

   It removes the requirement that lists are not printed if empty. Instead,
   lists are not printed if they do not have their default value.

   It is now clarified that the textual format
   `my-pass{string-list=""}` or `my-pass{string-list={}}`
   is interpreted as "empty list". This makes it imposssible to specify
   that ListOption `string-list` should be a size-1 list containing the
   empty string. However, `my-pass{string-list={"",""}}` *does* specify
   a size-2 list containing the empty string. This behavior seems preferable
   to allow for overriding non-empty defaults as described above.

christopherbate commented:
Note: this still doesn't handle empty string literals correctly. I'm not sure what to do there other than disallow having non-empty `llvm::list_init` for `ListOption<std::string>`

River707 reviewed (commented):
Will review in the next day, my brain needs to carefully read this given how finicky the parsing here can be,

River707 commented on mlir/lib/Pass/PassRegistry.cpp:
```diff
@@ -186,6 +186,27 @@ const PassPipelineInfo *mlir::PassPipelineInfo::lookup(StringRef pipelineArg) {
 // PassOptions
 //===----------------------------------------------------------------------===//
 
+static size_t findChar(StringRef str, size_t index, char c) {
```
Can you write a doc comment for this function?

River707 commented on mlir/test/lib/Pass/TestPassManager.cpp:
```diff
@@ -85,7 +85,9 @@ struct TestOptionsPass
 
-  void runOnOperation() final {}
+  void runOnOperation() final {
+    llvm::errs() << "string-list has size " << stringListOption.size() << "\n";
```
Leftover debugging?

christopherbate replied:
thanks, removed

River707 commented on mlir/test/Pass/pipeline-options-parsing.mlir:
```diff
@@ -27,3 +43,11 @@
+// CHECK_10: builtin.module(func.func(test-options-pass{enum=zero  string= string-list={,}}))
+
+
+func.func @func() {
+  return
+}
```
Is this function actually necessary?

christopherbate replied:
removed it, thanks

christopherbate committed:
[mlir] Attempt to resolve edge cases in PassPipeline textual format

This commit makes the following changes:

1. Previously certain pipeline options could cause the options parser to
   get stuck in an an infinite loop. An example is:

   ```
   mlir-opt %s -verify-each=false -pass-pipeline='builtin.module(func.func(test-options-super-pass{list={list=1,2},{list=3,4}}))''
   ```

   In this example, the 'list' option of the `test-options-super-pass`
   is itself a pass options specification (this capability was added in
   https://github.com/llvm/llvm-project/issues/101118).

   However, while the textual format allows `ListOption<int>` to be given
   as `list=1,2,3`, it did not allow the same format for
   `ListOption<T>` when T is a subclass of `PassOptions` without extra
   enclosing `{....}`. Lack of enclosing `{...}` would cause the infinite
   looping in the parser.

   This change resolves the parser bug and also allows omitting the
   outer `{...}` for `ListOption`-of-options.

2. Previously, if you specified a default list value for your
   `ListOption`, e.g. `ListOption<int> opt{*this, "list", llvm::list_init({1,2,3})}`,
   it would be impossible to override that default value of `{1,2,3}` with
   an *empty* list on the command line, since `my-pass{list=}` was not allowed.

   This was not allowed because of ambiguous handling of lists-of-strings
   (no literal marker is currently required).

   This change makes it explicit in the ListOption construction that we
   would like to treat all ListOption as having a default value of "empty"
   unless otherwise specified (e.g. using `llvm::list_init`).

   It removes the requirement that lists are not printed if empty. Instead,
   lists are not printed if they do not have their default value.

   It is now clarified that the textual format
   `my-pass{string-list=""}` or `my-pass{string-list={}}`
   is interpreted as "empty list". This makes it imposssible to specify
   that ListOption `string-list` should be a size-1 list containing the
   empty string. However, `my-pass{string-list={"",""}}` *does* specify
   a size-2 list containing the empty string. This behavior seems preferable
   to allow for overriding non-empty defaults as described above.

joker-eph commented:
Wouldn't this be: `my-pass{option=""}`?

j2kun commented:
@joker-eph that method still seems to fail. It seems that to define an empty string I have to actually use the `list={}` syntax, which is strange (it's even strange that `list={}` allows this given the option is not typed as a list).

See for example https://github.com/google/heir/commit/ccba176a7f2b29f226f9e032ec56fd86ba8fee64#diff-6f8c01b4d228f97d2a800c5b5deca07240dd1203a0063e775b6a9f8e49cc9b9c the file tests/Dialect/Polynomial/Transforms/elementwise_to_affine.mlir which corresponds to the option defined here: https://github.com/google/heir/blob/276bc7c53ad9b074ad3b8dde2756b1980e4c2707/lib/Transforms/ElementwiseToAffine/ElementwiseToAffine.td#L31

I am on vacation for a few weeks so I may not have time to repro this upstream.

christopherbate commented:
Sorry just saw the post-merge discussion. 

> This PR seems to change the behavior of a string option with no default, when passing the empty string my-pass{option=}. Is it expected that this means you cannot override an unset string option with an empty string?

I don't think it was intended that the syntax be changed for plain string options. An option of type `std::string` with no default has an empty string value, however (because it gets default-constructed).

> It seems that to define an empty string I have to actually use the list={} syntax, which is strange (it's even strange that list={} allows this given the option is not typed as a list).

The example you gave in your last message is a `ListOption`? Are you saying there's a problem with `Option<...., "std::string",...>` or `ListOption<..., "std::string",...>`?

j2kun commented:
@christopherbate indeed, the option is defined as

```
    ListOption<"convertOps","convert-ops", "std::string",
                "comma-separated list of ops to run this pass on ">,
```

I guess it's not so important to preserve that particular syntax, if an unset string option cannot be distinguished from an option set to empty string anyway. It's a bit weird that the syntax is invalid now, and also weird that setting it to an empty list (even though it's a string option) is valid.
